src/github.com/getlantern/flashlight/client/pluggable_transports.go

```diff
@@ -0,0 +1,86 @@
+package client
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+
+	"github.com/Yawning/obfs4/transports/obfs4"
+	"github.com/getlantern/keyman"
+	"github.com/getlantern/tlsdialer"
+
+	"git.torproject.org/pluggable-transports/goptlib.git"
+)
+
+type dialFN func() (net.Conn, error)
+type dialFactory func(*ChainedServerInfo, string) (dialFN, error)
+
+var pluggableTransports = map[string]dialFactory{
+	"":      defaultDialFactory,
+	"obfs4": obfs4DialFactory,
+}
+
+func defaultDialFactory(s *ChainedServerInfo, deviceID string) (dialFN, error) {
+	netd := &net.Dialer{Timeout: chainedDialTimeout}
+	forceProxy := ForceChainedProxyAddr != ""
+	addr := s.Addr
+	if forceProxy {
+		log.Errorf("Forcing proxying to server at %v instead of configured server at %v", ForceChainedProxyAddr, s.Addr)
+		addr = ForceChainedProxyAddr
+	}
+
+	if s.Cert == "" && !forceProxy {
+		log.Error("No Cert configured for chained server, will dial with plain tcp")
+		return func() (net.Conn, error) {
+			return netd.Dial("tcp", addr)
+		}, nil
+	}
+	log.Trace("Cert configured for chained server, will dial with tls over tcp")
+	cert, err := keyman.LoadCertificateFromPEMBytes([]byte(s.Cert))
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse certificate: %s", err)
+	}
+	x509cert := cert.X509()
+	sessionCache := tls.NewLRUClientSessionCache(1000)
+	return func() (net.Conn, error) {
+		conn, err := tlsdialer.DialWithDialer(netd, "tcp", addr, false, &tls.Config{
+			ClientSessionCache: sessionCache,
+			InsecureSkipVerify: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if !forceProxy && !conn.ConnectionState().PeerCertificates[0].Equal(x509cert) {
+			if err := conn.Close(); err != nil {
+				log.Debugf("Error closing chained server connection: %s", err)
+			}
+			return nil, fmt.Errorf("Server's certificate didn't match expected!")
+		}
+		return conn, err
+	}, nil
+}
+
+func obfs4DialFactory(s *ChainedServerInfo, deviceID string) (dialFN, error) {
+	if s.Cert == "" {
+		return nil, fmt.Errorf("No Cert configured for obfs4 server, will dial with plain tcp")
+	}
+
+	tr := obfs4.Transport{}
+	cf, err := tr.ClientFactory("")
```

Comment:
Maybe just a comment about why we pass the empty string here? Looks like that string's not even used on the obfs4 side @oxtoacart?